### PR TITLE
[Bifrost] Watch tail updates

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[Bifrost] Watch tail updates

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1743).
* #1757
* #1753
* #1747
* #1744
* __->__ #1743
* #1742

### CODE REVIEW

File: crates/bifrost/src/providers/local_loglet/mod.rs [hunk 1]
```diff
@@ -56,7 +57,8 @@ struct LocalLoglet {
     last_committed_offset: AtomicU64,
     next_write_offset: Mutex<LogletOffset>,
     sealed: AtomicBool,
-    release_watch: OffsetWatch,
+    // watches the tail state of ths loglet
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/local_loglet/mod.rs:60]: `s/ths/this/`

File: crates/bifrost/src/providers/local_loglet/read_stream.rs [hunk 2]
```diff
@@ -14,38 +14,41 @@ use std::sync::Arc;
...
+    last_known_tail: LogletOffset,
     /// Last offset to read before terminating the stream. None means "tailing" reader.
     read_to: Option<LogletOffset>,
     #[pin]
     iterator: DBRawIteratorWithThreadMode<'static, DB>,
     #[pin]
-    release_watch: WatchStream<LogletOffset>,
+    tail_watch: BoxStream<'static, TailState<LogletOffset>>,
     #[pin]
     terminated: bool,
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/local_loglet/read_stream.rs:48]: I think that these fields no longer need to be pinned. That way we could get completely rid of `pin_project` for this type.

File: crates/bifrost/src/providers/memory_loglet.rs [hunk 3]
```diff
@@ -179,9 +183,9 @@ struct MemoryReadStream {
     /// The next offset to read from
     read_pointer: LogletOffset,
     #[pin]
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/memory_loglet.rs:185]: I think there is no need for the pin projections here.

**Review by tillrohrmann — APPROVED** [review]
Nice work @AhmedSoliman :-) LGTM. +1 for merging. My comment about `pin-project` is unrelated.

File: crates/bifrost/src/providers/local_loglet/read_stream.rs (hunk 2 above)
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/local_loglet/read_stream.rs:48]: I'll do that in a follow up PR.

**Pull request merged**